rfcs/20201221-tfmot-compression-api.md

```diff
@@ -0,0 +1,455 @@
+Our API also provides guidelines for testing and benchmark. For now, we only have guidelines to apply a compression algorithm for simple MNIST vision cases. We'd like to provide an example for tensorflow [official models](https://github.com/tensorflow/models/tree/master/official) in the future.
+
+### Tutorials and Examples
+We provide the tutorial for [SVD](https://en.wikipedia.org/wiki/Singular_value_decomposition) compression algorithm that shows how we implement the SVD algorithm using TFMOT compression API by colab. This tutorial includes:
+
+#### Algorithm developer side
+1. The algorithm developer implementing the SVD algorithm uses the `WeightCompressor` class.
+
+```python
+class SVD(algorithm.WeightCompressor):
+  """SVD compression module config."""
+
+  def __init__(self, params):
+    self.params = params
+
+  def init_training_weights(
+      self, pretrained_weight: tf.Tensor):
+    """Init function from pre-trained model case."""
+    rank = self.params.rank
+
+    # Dense Layer
+    if len(pretrained_weight.shape) == 2:
+      u, sv = tf_svd_factorization_2d(pretrained_weight, rank)
+    else:
+      raise NotImplementedError('Only for dimension=2 is supported.')
+
+    self.add_training_weight(
+        name='u',
+        shape=u.shape,
+        dtype=u.dtype,
+        initializer=tf.keras.initializers.Constant(u))
+    self.add_training_weight(
+        name='sv',
+        shape=sv.shape,
+        dtype=sv.dtype,
+        initializer=tf.keras.initializers.Constant(sv))
+
+  def project_training_weights(self, u: tf.Tensor, sv: tf.Tensor) -> tf.Tensor:
+    return tf.matmul(u, sv)
+
+  def get_compressible_weights(
+      self, original_layer: tf.keras.layers.Layer) -> List[str]:
+    rank = self.params.rank
+    if isinstance(original_layer, tf.keras.layers.Dense):
+      input_dim = original_layer.kernel.shape[0]
+      output_dim = original_layer.kernel.shape[1]
+      if input_dim * output_dim > (input_dim + output_dim) * rank:
+        return ['kernel']
+    return []
+```
+
+2. Export the model developer API for the SVD algorithm.
+```python
+class SVDParams(object):
+  """Define container for parameters for SVD algorithm."""
+
+  def __init__(self, rank):
+    self.rank = rank
+
+def optimize(to_optimize: tf.keras.Model, params: SVDParams) -> tf.keras.Model:
+  """Model developer API for optimizing a model."""
+
+  def _optimize_layer(layer):
+    # Require layer to be built so that the SVD-factorized weights
+    # can be initialized from the weights.
+    if not layer.built:
+      raise ValueError(
+          'Applying SVD currently requires passing in a built model')
+
+    return algorithm.create_layer_for_training(layer, algorithm=SVD(params))
+
+  return tf.keras.models.clone_model(
+      to_optimize, clone_function=_optimize_layer)
+```
+
+#### Model developer side
+1. The model developer uses the SVD algorithm.
+```python
+params = SVDParams(rank=32)
```

Comment:
This object seems like an implementation detail. And it isn't clear why `optimize` is separate from `SVD`.

We could simply do:

```python
compressed_model = SVD(rank=32).optimize(model)
```

and have `optimize` be a method on the `SVD` class. The `SVD` class would contain the weights -- no need for a separate container object.

Reply:
Okay, That makes more user experience simpler.

I thought params & optimize is not a part of core API. because it can be different for each algorithms.
But it's okay to make API shape like that on SVD example.

SVD only has one optimize methods, but other algorithms like pruning can have two steps.
(1. original -> training model (mask + weight), 2. training model -> compressible model(sparse tensor))

I've updated RFCs to apply that. Thanks!